### PR TITLE
Import Gecko as a single module [gecko-integration2-dev]

### PR DESCRIPTION
**Pull request by jakubcerveny**
Gecko has recently been released under the BSD license (https://github.com/LLNL/gecko).

This PR imports the library into MFEM as a single module. It's up for discussion if this is the way we want to do it.

The interface in Mesh has been improved and better documented.

TODO: 
* [x] Move the demo from `ex1` to `mesh-explorer`
* [x] Update makefile and CMake
* [x] Update CHANGELOG

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1282](https://github.com/mfem/mfem/pull/1282) | @jakubcerveny | @tzanio | @lindstro + @tzanio | 02/16/20 | 03/14/20 | ⌛due 03/21/20 | |

### CODE REVIEW

**Comment by jakubcerveny** [comment]
I moved the demo code to `mesh-explorer`. Example for 3x refined `star.mesh`:

Initial ordering just after refinement (note that this used to be much worse before #1119):

no-order

After Gecko ordering with default parameters:

gecko-order

**Comment by jakubcerveny** [comment]
(Just for comparison, this is the other option already in `mesh-explorer`, the Hilbert sort:)

hilbert-order

**Comment by tzanio** [comment]
Please address Appveyor and Travis failures.

**Comment by jakubcerveny** [comment]
Question: what is the documentation check? Can we disable it specifically for `gecko.?pp`?

**Comment by tzanio** [comment]
> @tzanio Question: what is the documentation check? Can we disable it specifically for `gecko.?pp`?

It is https://github.com/mfem/mfem/blob/master/tests/scripts/documentation, basically running Doxygen. We can exclude the Gecko files in https://github.com/mfem/mfem/blob/master/doc/CodeDocumentation.conf.in if that's the appropriate thing to do...

**Comment by jakubcerveny** [comment]
This should now be OK, the branch-history check is a false positive...

**Comment by lindstro** [comment]
@jakubcerveny Thanks for integrating gecko into MFEM.  For my benefit, why does the code not compile with GECKO_WITH_NONRECURSIVE enabled?  It works fine for me on Linux, Mac, and Windows (outside of MFEM).  Is there a Travis or AppVeyor build that I should be looking at?  If there's an issue with gecko, I'd like to fix it.

I also fixed a number of Windows compiler warnings on the `master` branch after the 1.0.0 release.  Not sure what gecko commit you're working off of.

**Comment by jakubcerveny** [comment]
@lindstro This PR is based on what currently is in master in https://github.com/LLNL/gecko/commits/master, so we should be up to date. 

The problem with the nonrecursive version of `Subgraph::optimize` was that it has no arguments, but there is no such declaration in `class Subgraph`. It wouldn't help anyway, because the outer method `Subgraph::optimize(uint p)` calls `optimize(0, n)`. I wasn't sure how to fix that so I just removed the #ifdef (because Doxygen tends to go inside disabled code and produces warnings).

**Comment by lindstro** [comment]
I fixed the nonrecursive signature and a number of other minor issues that were exposed when testing with other compile-time options enabled.  I'm not sure MFEM benefits from these options, however.

**Comment by tzanio** [comment]
@lindstro and @rw-anderson, do you need more time to review this?

File: general/gecko.cpp [hunk 1]
```diff
@@ -0,0 +1,1272 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
```
Comment (tzanio) [inline general/gecko.cpp]: Use the new Copyright header

**Comment by jakubcerveny** [comment]
Yes, we expect very few changes in the upstream code. Should there be more activity, we may try to automate the import process.

File: general/gecko.cpp (hunk 1 above)
Comment (lindstro) [reply general/gecko.cpp]: The gecko part of the header also needs updating.  In particular, it currently includes the old LLNL-CODE release number.  See [gecko.h](https://github.com/LLNL/gecko/blob/master/include/gecko.h).

**Comment by tzanio** [comment]
@rw-anderson, can you please review, this is overdue...

File: general/gecko.cpp (hunk 1 above)
Comment (jakubcerveny) [reply general/gecko.cpp]: I have included the following block in `gecko.{hpp,cpp}` (note the change to "see the above license"):
```
// Copyright (c) 2019-2020, Lawrence Livermore National Security, LLC and other
// gecko project contributors. See the above license for details.
// SPDX-License-Identifier: BSD-3-Clause
```
However, I don't see any LLNL-CODE number (neither in our `gecko.{hpp,cpp}` nor in `gecko.h`)

Comment (lindstro) [reply general/gecko.cpp]: The (new) LLNL-CODE number is in README.md.  I just wanted the old one taken out.

Comment (lindstro) [reply general/gecko.cpp]: And in case you want to include it, it's LLNL-CODE-800597.

Comment (jakubcerveny) [reply general/gecko.cpp]: Gecko LLNL-CODE is now included in the headers in `gecko.{hpp,cpp}`, thanks.

File: CHANGELOG [hunk 2]
```diff
@@ -102,6 +102,11 @@ Meshing improvements
 
 - The TMOP algorithms were also improved to support non-conforming AMR meshes.
 
+- The graph linear ordering library Gecko, previously an external dependency, is
+  now included directly in MFEM. As a result, the method
+  Mesh::GetGeckoElementOrdering is now always available. The interface has also
+  been improved, see the mesh-explorer miniapp for an example of use.
+
```
Comment (tzanio) [inline CHANGELOG]: Can you merge master in and move this to the new 4.1.1 section?

File: CHANGELOG [hunk 3]
```diff
@@ -102,6 +102,11 @@ Meshing improvements
 
 - The TMOP algorithms were also improved to support non-conforming AMR meshes.
 
+- The graph linear ordering library Gecko, previously an external dependency, is
+  now included directly in MFEM. As a result, the method
+  Mesh::GetGeckoElementOrdering is now always available. The interface has also
+  been improved, see the mesh-explorer miniapp for an example of use.
```
Comment (tzanio) [inline CHANGELOG]:
```suggestion
  now included directly in MFEM. As a result, Mesh::GetGeckoElementOrdering is
  always available. The interface has also been improved, see for example the
  mesh-explorer miniapp.
```

File: general/gecko.hpp [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_GECKO_HPP
+#define MFEM_GECKO_HPP
+
+// This file collects the sources of the Gecko library as a single module.
+// The original library can be found at https://github.com/LLNL/gecko
+// Used here with permission.
+
```
Comment (tzanio) [inline general/gecko.hpp]: Mention Gecko at the end of the top-level `NOTICE` file

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Nightly regression test errors:

```
In file included from general/gecko.cpp:12:0:
general/gecko.hpp:58:1: warning: "/*" within comment [-Wcomment]
 /* ------------------------------------------------------------------------------
 ^
In file included from mesh/mesh.cpp:21:0:
mesh/../general/gecko.hpp:58:1: warning: "/*" within comment [-Wcomment]
 /* ------------------------------------------------------------------------------
 ^
```

**Comment by jakubcerveny** [comment]
Should be fixed, thanks.

I'm not sure if I can push to next and/or elsewhere...

Jakub

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by camierjs** [comment]
We have a lot of new warnings on `Lassen` with the `cuda` build:
```
general/gecko.hpp(341): warning: overloaded virtual function "Gecko::Functional::sum" is only partially overridden in class "Gecko::FunctionalHarmonic"
general/gecko.hpp(368): warning: overloaded virtual function "Gecko::Functional::sum" is only partially overridden in class "Gecko::FunctionalGeometric"
general/gecko.hpp(390): warning: overloaded virtual function "Gecko::Functional::sum" is only partially overridden in class "Gecko::FunctionalSMR"
general/gecko.hpp(412): warning: overloaded virtual function "Gecko::Functional::sum" is only partially overridden in class "Gecko::FunctionalArithmetic"
general/gecko.hpp(473): warning: overloaded virtual function "Gecko::Functional::sum" is only partially overridden in class "Gecko::FunctionalRMS"
general/gecko.hpp(503): warning: overloaded virtual function "Gecko::Functional::sum" is only partially overridden in class "Gecko::FunctionalMaximum"
general/gecko.hpp(503): warning: overloaded virtual function "Gecko::Functional::accumulate" is only partially overridden in class "Gecko::FunctionalMaximum"
```
@adrienbernede started to look at adding one test that could catch these to help us.

**Comment by lindstro** [comment]
The warnings are valid though inconsequential, I believe.  I've pushed a fix to the gecko repo that hopefully addresses this.

**Comment by tzanio** [comment]
@jakubcerveny, do you want to add this fix

**Comment by jakubcerveny** [comment]
Added in #1372

**Comment by tzanio** [comment]
Thank you!

**Pull request merged**